pkg/registries/docker/docker.go

```diff
@@ -86,34 +86,39 @@ type Config struct {
 	TokenSource oauth2.TokenSource
 }
 
-// NewDockerRegistryWithConfig creates a new instantiation of the docker registry
-// TODO(cgorman) AP-386 - properly put the base docker registry into another pkg
-func NewDockerRegistryWithConfig(cfg Config, integration *storage.ImageIntegration) (*Registry, error) {
-	endpoint := cfg.Endpoint
+func (c *Config) formatURL() string {
+	endpoint := c.Endpoint
 	if strings.EqualFold(endpoint, "https://docker.io") || strings.EqualFold(endpoint, "docker.io") {
 		endpoint = "https://registry-1.docker.io"
 	}
-	url := urlfmt.FormatURL(endpoint, urlfmt.HTTPS, urlfmt.NoTrailingSlash)
+	return urlfmt.FormatURL(endpoint, urlfmt.HTTPS, urlfmt.NoTrailingSlash)
+}
 
-	// if the registryServer endpoint contains docker.io then the image will be docker.io/namespace/repo:tag
-	registryServer := urlfmt.GetServerFromURL(url)
-	if strings.Contains(cfg.Endpoint, "docker.io") {
-		registryServer = "docker.io"
-	}
-	var transport http.RoundTripper
-	if cfg.Insecure {
+// Transport returns a transport that provides authentication for the Docker registry.
+func (c *Config) Transport() registry.Transport {
+	transport := proxy.RoundTripper()
+	if c.Insecure {
 		transport = proxy.RoundTripperWithTLSConfig(&tls.Config{
 			InsecureSkipVerify: true,
 		})
-	} else {
-		transport = proxy.RoundTripper()
 	}
-	if cfg.TokenSource != nil {
-		transport = &oauth2.Transport{Base: transport, Source: cfg.TokenSource}
+	if c.TokenSource != nil {
+		transport = &oauth2.Transport{Base: transport, Source: c.TokenSource}
+	}
+	return registry.WrapTransport(transport, strings.TrimSuffix(c.formatURL(), "/"), c.Username, c.Password)
+}
+
+// NewDockerRegistryWithConfig creates a new instantiation of the docker registry
+// TODO(cgorman) AP-386 - properly put the base docker registry into another pkg
+func NewDockerRegistryWithConfig(cfg Config, integration *storage.ImageIntegration, transport registry.Transport) (*Registry, error) {
```

Comment:
Hmm, seeing the current usages of this, it might be non-trivial what the transport should be.

Also, for most use-cases it seems like the transport _usually_ is encoded within the `cfg`, so it might be worthwhile to do the following:
- have field `transport` in the config.
- in case `transport` is empty / not set, fallback to the existing logic of `config.Transport()`.
- for use-cases where a custom transport has to be used, set the `transport` field explicitly.

This way it'll be a cleaner interface and you'll only need to pass in the config, in case you want to use the default transport you don't set the field, in case of customizations you do. This again also minimizes the changes required for callers as the arguments will stay the same.